src/sync/channel.zig

```diff
@@ -1,170 +1,455 @@
+            block: Atomic(?*Buffer),
 
-        pub fn send(self: *Self, value: T) error{ OutOfMemory, ChannelClosed }!void {
-            if (self.closed.load(.monotonic)) {
-                return error.ChannelClosed;
+            fn deinit(pos: *Position, allocator: Allocator) void {
+                if (pos.block.load(.monotonic)) |block| {
+                    block.deinit(allocator);
+                    allocator.destroy(block);
+                }
             }
-            var buffer_lock = self.buffer.lock();
-            defer buffer_lock.unlock();
+        };
 
-            var buffer: *std.ArrayList(T) = buffer_lock.mut();
-            try buffer.append(value);
+        const Buffer = struct {
+            next: Atomic(?*Buffer),
+            slots: [BLOCK_CAP]Slot,
 
-            self.has_value.signal();
-        }
-
-        pub fn sendBatch(self: *Self, value: std.ArrayList(T)) error{ OutOfMemory, ChannelClosed }!void {
-            if (self.closed.load(.monotonic)) {
-                return error.ChannelClosed;
+            fn create(allocator: Allocator) !*Buffer {
+                const new = try allocator.create(Buffer);
+                @memset(&new.slots, Slot.uninit);
+                new.next = Atomic(?*Buffer).init(null);
+                return new;
             }
-            var buffer_lock = self.buffer.lock();
-            defer buffer_lock.unlock();
 
-            var buffer: *std.ArrayList(T) = buffer_lock.mut();
-            try buffer.appendSlice(value.items);
+            fn destroy(block: *Buffer, start: usize, allocator: Allocator) void {
+                for (start..BLOCK_CAP - 1) |i| {
+                    const slot = &block.slots[i];
 
-            self.has_value.signal();
-        }
+                    if (slot.state.load(.acquire) & READ_FROM == 0 and
+                        slot.state.fetchOr(DESTROYED, .acq_rel) & READ_FROM == 0)
+                    {
+                        return;
+                    }
+                }
 
-        pub fn len(self: *Self) error{ChannelClosed}!u64 {
-            if (self.closed.load(.monotonic)) {
-                return error.ChannelClosed;
+                allocator.destroy(block);
             }
-            var buffer = self.buffer.lock();
-            defer buffer.unlock();
 
-            return buffer.get().items.len;
+            fn deinit(block: *Buffer, allocator: Allocator) void {
+                if (block.next.load(.monotonic)) |n| {
+                    n.deinit(allocator);
+                    allocator.destroy(n);
+                }
+            }
+        };
+
+        const Slot = struct {
+            value: T,
+            state: Atomic(usize),
+
+            const uninit: Slot = .{
+                .value = undefined,
+                .state = Atomic(usize).init(0),
+            };
+        };
+
+        pub fn init(allocator: Allocator, initial_capacity: usize) !Self {
+            _ = initial_capacity; // TODO: do something with this
```

Comment:
do we want to fix this todo?

Reply:
Hmm, I think I'd like to leave it for now. It's not necessary for any real performance reasons for now, however, it might be useful to add if we realize that the channel is slowing us down at some point.

Reply:
ok - maybe we just remove the parameter for now and add a comment saying we could add support for initial capacity if performance is slow